lib/puppet-lint/plugins/check-params_empty_string_assignment.rb

```diff
@@ -12,7 +12,7 @@ def check
         next if type.empty?
         next if default_value.empty?
         # if the parameter has datatype String/String[] and a value of ''
-        if type[0].type == :TYPE && type[0].value =~ /^String/ && default_value[0].value == ''
+        if type[0].type == :TYPE && type[0].value =~ /^(String|Variant)/ && default_value[0].value == ''
```

Comment:
I feel like at this point it would be useful to have even richer parsing so you don't manually need to parse `String[0]` into `String` and `0`. Perhaps worth an issue to `puppet-lint` itself?

Reply:
that would help if we differentiate between `String` and `String[0]`, but that's currently not the case. Instead of the regex I could do an explicit match against all types, but I think this is sufficient (because we don't need to differentiate between the non-allowed types)